rep-2011.rst

```diff
@@ -0,0 +1,1421 @@
+
+When creating a publisher or subscription, the caller normally provides:
+
+- a topic name,
+- QoS settings, and
+- a topic type
+
+Where the topic type is represented as a string and is automatically deduced based on the type given to the function that creates the entity.
+The type may be passed as a template parameter in C++ or as an argument to the function in Python.
+
+For example, creating a publisher for the C++ type ``std_msgs::msg::String`` using ``rclcpp`` may result in a topic type like the string ``std_msgs/msg/String``.
+
+All of the above items are used by the middleware to determine if two endpoints should communicate or not, and this REP proposes that the type version be added to this list of provided information.
+
+Nothing needs to change from the user's perspective, as the type version can be extracted automatically based on the topic type given, either at the ``rcl`` layer or in the ``rmw`` implementation itself.
+That is to say, how users create things like publishers and subscription should not need to change, no matter which programming language is being used.
+
+However, the type version hash would become something that the ``rmw`` implementation is provided and aware of in the course of creating a publisher or subscription.
+The decision of whether or not to use that information to enforce type compatibility would be left to the middleware, rather than implementing it as logic in ``rcl`` or other packages above the ``rmw`` API.
+
+The method for implementing the detection and enforcement of type version mismatches is left up to the middleware.
+Some middlewares will have tools to handle this without this new type version hash and others will implement something like what would be possible in the ``rcl`` and above layers using the type version hash.
+By keeping this a detail of the ``rmw`` implementation, we allow the ``rmw`` implementations to make optimizations where they can.
+
+Recommended Strategy for Enforcing that Type Versions Match
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+If the middleware has a feature to handle type compatibility already, as is the case with DDS-XTypes which is discussed later, then that can be used to enforce type safety, and then the type version hash would only be used for debugging and for storing in recordings.
+
+However, if the middleware lacks this kind of feature, then the recommended strategy for accomplishing this in the ``rmw`` implementation is to simply concatenate the type name and the type version hash with double underscores and then use that as the type name given to the underlying middleware.
+For example, a type name using this approach may look like this:
+
+.. code::
+
+    sensor_msgs/msg/Image__RIHS1_XXXXXXXXXXXXXXXXXXXX
```

Comment:
Small concern here, maybe just due to my bad memory - do we require all RMW impls to use the provided QoS profiles? If so, and we are planning to add `USER_DATA` to provide type hash, then do we need this backup strategy of concatenated type name?

If USER_DATA not universally viable - if one does use this strategy, is the plaintext typename totally redundant (because type name is in the TypeDescription object that produces the hash, and so prevents hash collision for types even of the same layout with different names)? If redundant, would it be better to just advertise type `RIHS_1_xxx`? Maybe not, since hash is not reversible, and the prefix allows for checking for transfer functions. I may have not been thinking of the whole scope of the REP.

Reply:
Not sure about QoS profiles.

But for advertised type names.. Putting the type name as a prefix in the advertised type will allow us to bypass querying for the TypeDescription object in the case where we just want the name of the type the participant is trying to use (e.g. in a tool to see what participants are trying to communicate using the same type (the prefix), but can't because of the layout (different hashes))